Use a single peer state map for all channel phases in peer state

dunxen opened the pull request:
Fixes #2422

dunxen commented:
> Was nice while it lasted, but, yea, I think this makes sense 😢

Yeah, there were *some* benefits to the multi-map approach, but I believe there's just too many negatives and harder to consolidate. The one con to single map is that there is added verbosity in places but I think we can `impl` helper methods on a newtype for the map to make things more readable and dedup stuff.

alecchendev commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2293,8 +2295,14 @@ where
+						_ => None,
+					})
+					.filter(f)
+				{
 					let details = ChannelDetails::from_channel_context(&channel.context, best_block_height,
 						peer_state.latest_features.clone(), &self.fee_estimator);
```
nit: could probably just add this as another `.map` and then `res.extend` the whole iterator (and similarly in `list_channels`)

vladimirfomene reviewed (commented):
Just a few questions

alecchendev reviewed (commented):
Got maybe ~half way through

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1001,8 +985,7 @@ where
 	/// expose them to users via a PaymentClaimable event. HTLCs which do not meet the requirements
-	/// here are failed when we process them as pending-forwardable-HTLCs, and entries are removed
-	/// after we generate a PaymentClaimable upon receipt of all MPP parts or when they time out.
+	/// here are failed when we process them as pending-forwardable-HTLCs, and entries are  after we generate a PaymentClaimable upon receipt of all MPP parts or when they time out.
```
oops looks like removed was removed. Unintended change

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1789,64 +1772,71 @@ macro_rules! convert_chan_err {
+			},
+			ChannelError::Close(msg) => {
+				log_error!($self.logger, "Closing channel {} due to close-required error: {}", log_bytes!($channel_id[..]), msg);
+				match $channel_phase {
+					ChannelPhase::Funded(chan) => {
+						update_maps_on_chan_removal!($self, chan.context);
+						let shutdown_res = chan.context.force_shutdown(true);
+						(true, MsgHandleErrInternal::from_finish_shutdown(msg, *$channel_id, chan.context.get_user_id(),
+							shutdown_res, $self.get_channel_update_for_broadcast(&chan).ok(), chan.context.get_value_satoshis()))
+					},
+					_ => {
+						let chan_context = $channel_phase.chan_context_mut();
+						update_maps_on_chan_removal!($self, chan_context);
+						let shutdown_res = chan_context.force_shutdown(true);
+						(true, MsgHandleErrInternal::from_finish_shutdown(msg, *$channel_id, chan_context.get_user_id(),
+							shutdown_res, None, chan_context.get_value_satoshis()))
+					}
```
Yip let's do that now :)

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1789,64 +1772,71 @@ macro_rules! convert_chan_err {
 			Ok(res) => res,
 			Err(e) => {
-				let (drop, res) = convert_chan_err!($self, e, $entry.get_mut().context, $entry.key(), UNFUNDED);
+				let key = $entry.key().clone();
```
Yeah they're pretty cheap but I'll check which are all unnecessary and remove them!

alecchendev replied:
Oh woops, true

dunxen replied:
Thanks though, looks better :)

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2590,10 +2673,19 @@ where
 			};
 			if let hash_map::Entry::Occupied(chan) = peer_state.channel_by_id.entry(channel_id.clone()) {
```
Thanks yeah there are a few other naming inconsistencies I see. Will try get them all :+1:

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3928,62 +3955,64 @@ where
-												log_trace!(self.logger, "Failed to forward HTLC with payment_hash {}: {}", log_bytes!(payment_hash.0), msg);
-											} else {
-												panic!("Stated return value requirements in send_htlc() were not met");
+						hash_map::Entry::Occupied(mut phase) => {
+							if let ChannelPhase::Funded(chan) = phase.get_mut() {
```
Yup, we're not doing anything fundamental with the entry other than getting a mutable ref to the phase, so let's do that.

alecchendev commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9101,27 +9253,36 @@ where
-						log_error!(args.logger, "Got duplicate outbound SCID alias; {}", chan.context.outbound_scid_alias());
-						return Err(DecodeError::InvalidValue);
+			for (chan_id, phase) in peer_state.channel_by_id.iter_mut() {
+				match phase {
+					ChannelPhase::Funded(chan) => {
+						if chan.context.outbound_scid_alias() == 0 {
```
Also here too I think

alecchendev commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -8089,31 +8234,30 @@ where
 		let mut serializable_peer_count: u64 = 0;
 		{
 			let per_peer_state = self.per_peer_state.read().unwrap();
-			let mut unfunded_channels = 0;
-			let mut number_of_channels = 0;
+			let mut number_of_funded_channels = 0;
```
I like the clean up here :)

alecchendev commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -5784,22 +5864,26 @@ where
-							node_id: counterparty_node_id.clone(),
-							msg,
-						});
+				hash_map::Entry::Occupied(mut chan_phase_entry) => {
+					if let ChannelPhase::Funded(chan) = chan_phase_entry.get_mut() {
+						let (closing_signed, tx) = try_chan_phase_entry!(self, chan.closing_signed(&self.fee_estimator, &msg), chan_phase_entry);
```
> but I think we can impl helper methods on a newtype for the map to make things more readable and dedup stuff.

One thing I noticed while reviewing is there are a good amount of places where we want to use the `chan_phase_entry` and the `chan`, and only match on one variant, and otherwise if it's vacant in the hashmap or a different variant, we return roughly the same error. This might be a pattern where a helper/macro could simplify some things, not sure if it's best fit for this PR or a followup but just an idea :)

dunxen replied:
Keeping this for a followup for now, but might tack it on in this PR still.

alecchendev reviewed (commented):
Mainly was trying to look at where we match every variant vs. `_` (or do an `if let`) since one of the nice things with this PR is to have the compiler notify us of important places when adding new `ChannelPhase`s, and these places seem to be good (or at least lean towards being stricter) AFAICT, might have to think on these more though. Otherwise, mainly just some smaller style comments and whatnot

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -6709,24 +6842,28 @@ where
 				}
 
-				if let hash_map::Entry::Occupied(mut chan) = peer_state.channel_by_id.entry(channel_funding_outpoint.to_channel_id()) {
-					debug_assert_eq!(chan.get().context.get_funding_txo().unwrap(), channel_funding_outpoint);
-					if let Some((monitor_update, further_update_exists)) = chan.get_mut().unblock_next_blocked_monitor_update() {
-						log_debug!(self.logger, "Unlocking monitor updating for channel {} and updating monitor",
-							log_bytes!(&channel_funding_outpoint.to_channel_id()[..]));
-						if let Err(e) = handle_new_monitor_update!(self, channel_funding_outpoint, monitor_update,
-							peer_state_lck, peer_state, per_peer_state, chan)
-						{
-							errors.push((e, counterparty_node_id));
-						}
-						if further_update_exists {
-							// If there are more `ChannelMonitorUpdate`s to process, restart at the
-							// top of the loop.
-							continue;
+				if let hash_map::Entry::Occupied(mut chan_phase_entry) = peer_state.channel_by_id.entry(channel_funding_outpoint.to_channel_id()) {
+					if let ChannelPhase::Funded(chan) = chan_phase_entry.get_mut() {
+						debug_assert_eq!(chan.context.get_funding_txo().unwrap(), channel_funding_outpoint);
+						if let Some((monitor_update, further_update_exists)) = chan.unblock_next_blocked_monitor_update() {
+							log_debug!(self.logger, "Unlocking monitor updating for channel {} and updating monitor",
+								log_bytes!(&channel_funding_outpoint.to_channel_id()[..]));
+							if let Err(e) = handle_new_monitor_update!(self, channel_funding_outpoint, monitor_update,
+								peer_state_lck, peer_state, per_peer_state, chan_phase_entry)
+							{
+								errors.push((e, counterparty_node_id));
+							}
+							if further_update_exists {
+								// If there are more `ChannelMonitorUpdate`s to process, restart at the
+								// top of the loop.
+								continue;
+							}
+						} else {
+							log_trace!(self.logger, "Unlocked monitor updating for channel {} without monitors to update",
+								log_bytes!(&channel_funding_outpoint.to_channel_id()[..]));
 						}
 					} else {
-						log_trace!(self.logger, "Unlocked monitor updating for channel {} without monitors to update",
-							log_bytes!(&channel_funding_outpoint.to_channel_id()[..]));
+						//
```
no comment? lol

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7331,23 +7474,27 @@ where
+					let chan_context = match phase {
+						ChannelPhase::Funded(chan) => {
+							chan.remove_uncommitted_htlcs_and_mark_paused(&self.logger);
+							// We only retain retain unfunded channels that are not shutdown.
```
Oh wow I saw this too and thought I fixed it haha

alecchendev commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -5611,22 +5676,25 @@ where
-				// with the funding_signed so the channel can never go on chain).
-				if let Err(MsgHandleErrInternal { shutdown_finish: Some((res, _)), .. }) = &mut res {
-					res.0 = None;
+				if let ChannelPhase::Funded(chan) = e.insert(ChannelPhase::Funded(chan)) {
+					let mut res = handle_new_monitor_update!(self, monitor_res, peer_state_lock, peer_state,
+						per_peer_state, chan, MANUALLY_REMOVING_INITIAL_MONITOR,
+						{ peer_state.channel_by_id.remove(&new_channel_id) });
```
I think it should still be fine here since extracting `chan` shouldn't make much of a difference to that line?

alecchendev replied:
Oh wait I see what you mean, like `chan` may be mutably borrowed twice because of the remove - though I guess we don't try to use `chan` after the remove because locally extracting it works fine

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -5718,48 +5799,47 @@ where
+						if let Some(monitor_update) = monitor_update_opt {
+							break handle_new_monitor_update!(self, funding_txo_opt.unwrap(), monitor_update,
+								peer_state_lock, peer_state, per_peer_state, chan_phase_entry).map(|_| ());
+						}
+						break Ok(());
+					},
+					ChannelPhase::UnfundedInboundV1(_) | ChannelPhase::UnfundedOutboundV1(_) => {
+						let context = phase.chan_context_mut();
```
Yeah that was a mistake

dunxen commented:
Thanks for the review. Addressing.

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -8193,31 +8426,30 @@ where
-					}
+				for channel in peer_state.channel_by_id.iter().filter_map(
+					|(_, phase)| if let ChannelPhase::Funded(channel) = phase {
+						if channel.context.is_funding_initiated() { Some(channel) } else { None }
```
There's a case where we're the channel opener and the unfunded channel becomes a funded `Channel` when we generate the funding transaction and before we receive `funding_signed`. Then `is_funding_initiated()` would return false.

I think you had suggested we only promote a channel to funded when we receive funding_signed and create a monitor for it, but the consensus was that we do it when we generate the txid-based channel ID so we don't mix them in the maps (and some other stuff). But now we're back to having one map so they would be mixed anyway. So maybe we go back to that and then can get rid of `is_funding_initiated()`?

dunxen replied:
Yeah I wasn't thinking of changing it here. If we do end up shifting it then it'll be a separate PR after dual funding is considered. Focus is to get this PR through and get that V2 establishment PR ready for review and make progress there.

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -8665,14 +8897,14 @@ where
 					if channel.context.is_funding_initiated() {
```
See above.

dunxen replied:
Oh yes right, when we read this becomes unnecessary. I'll remove it in the same PR as the above change then it'll all be nicely contained.

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1811,32 +1811,76 @@ macro_rules! convert_chan_err {
+}
+
+/// Returns (boolean indicating if we should remove the Channel object from memory, a mapped error)
+macro_rules! convert_chan_phase_err {
```
The unfunded chan stuff gets removed in the next commit when the unfunded maps are removed. So I'm just basically working on merging `convert_funded_chan_err` and `convert_chan_phase_err` now.

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2459,37 +2457,49 @@ where
+								break Ok(());
+							},
+							_ => {
+								// If we reach this point, it means that the channel_id either refers to an unfunded channel or
```
Oh right. I missed a couple things down there to clean up too as pointed out by Wilmer.

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3557,8 +3468,8 @@ where
 		let peer_state = &mut *peer_state_lock;
-		let (chan, msg) = match peer_state.outbound_v1_channel_by_id.remove(temporary_channel_id) {
-			Some(chan) => {
+		let (chan, msg) = match peer_state.channel_by_id.remove(temporary_channel_id) {
```
Unfortunately we must take ownership so that `chan.get_funding_created` can receive `self`. Otherwise we can't get to the context to move it. There's also no Default implementation of `ChannelContext` to do a `mem::take` to take it through a `&mut self` when we construct the `Channel`. Creating such a default implementation doesn't seem practical. Also, creating another layer of Option for `context` in a channel would make things messy.

Not sure what's best here. The part I probably like least is that we need to re-add channels that are not `OutboundV1Channel`s.

dunxen replied:
Larger is ok and it's not that much larger. I was being really silly. `entry` works :100:

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2961,7 +2948,13 @@ where
+				let chan = match peer_state.channel_by_id.get_mut(&forwarding_id).map(
+					|chan_phase| if let ChannelPhase::Funded(chan) = chan_phase { Some(chan) } else { None }
+				).flatten() {
+				// let opt_chan_phase = peer_state.channel_by_id.get_mut(&forwarding_id);
+				// let chan = match opt_chan_phase.iter().find_map(
+				// 	|chan_phase| if let ChannelPhase::Funded(chan) = chan_phase { Some(chan) } else { None }
+				// ) {
```
When will I learn

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2028,7 +2008,18 @@ macro_rules! handle_new_monitor_update {
+		} else {
+			// We're not supposed to handle monitor updates for unfunded channels (they have no monitors to
+			// update).
+			debug_assert!(true);
```
:man_facepalming:

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2033,7 +2076,18 @@ macro_rules! handle_new_monitor_update {
-		handle_new_monitor_update!($self, $update_res, $peer_state_lock, $peer_state, $per_peer_state_lock, $chan_entry.get_mut(), MANUALLY_REMOVING_INITIAL_MONITOR, $chan_entry.remove_entry())
+		if let ChannelPhase::Funded(chan) = $chan_entry.get_mut() {
+			handle_new_monitor_update!($self, $update_res, $peer_state_lock, $peer_state, $per_peer_state_lock, chan, MANUALLY_REMOVING_INITIAL_MONITOR, { 
+				if let ChannelPhase::Funded(chan) = $chan_entry.remove() { chan } else { unreachable!() }
```
Oh sorry, I also mistakenly marked your previous feedback on this as resolved, but it seems you're right!

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2472,37 +2538,40 @@ where
 				let peer_state = &mut *peer_state_lock;
 
 				match peer_state.channel_by_id.entry(channel_id.clone()) {
-					hash_map::Entry::Occupied(mut chan_entry) => {
```
It could be set to `Err` by `break handle_new_monitor_update!(...);`

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -5902,37 +5954,40 @@ where
+					};
+					try_chan_phase_entry!(self, chan.update_add_htlc(&msg, pending_forward_info, create_pending_htlc_status, &self.fee_estimator, &self.logger), chan_phase_entry);
+				} else {
+					return Err(MsgHandleErrInternal::send_err_msg_no_close(format!("Got an update_add_htlc message for an unfunded channel!"), msg.channel_id));
```
Missed this one, sorry

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -5733,38 +5796,52 @@ where
-				Ok(())
+					Ok(())
+				} else {
+					debug_assert!(false);
```
Yeah that's dumb lol

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -5733,38 +5796,52 @@ where
+					Ok(())
+				} else {
+					debug_assert!(false);
+					let context = chan_phase_entry.get_mut().context_mut();
+					let channel_id = context.channel_id();
+					let user_id = context.get_user_id();
+					let shutdown_res = context.force_shutdown(false);
+					let channel_capacity = context.get_value_satoshis();
+					chan_phase_entry.remove();
+					Err(MsgHandleErrInternal::from_finish_shutdown("Got a channel_ready message for an unfunded channel!".to_owned(), channel_id,
+						user_id, shutdown_res, None, channel_capacity))
```
Right, I also think I may need to add issue_channel_close_events to convert_chan_phase_err. Or it was missing somewhere but I'll have another look. Will change.

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3468,40 +3472,60 @@ where
+		match peer_state.channel_by_id.entry(*temporary_channel_id) {
+			hash_map::Entry::Occupied(chan_phase_entry) => {
+				if let ChannelPhase::UnfundedOutboundV1(_) = chan_phase_entry.get() {
+					// We need to take ownership of the entry and promote the unfunded channel to a `Channel`,
```
Yeah that was only one lookup. There's no way in that case to avoid the removal and re-add of a channel of different type. I wish there was.

But 1 lookup vs 2 in happy path seems like a good enough reason to revert. So I'll do so.

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2057,7 +2113,21 @@ macro_rules! handle_new_monitor_update {
+			// update).
+			debug_assert!(false);
+			let channel_id = *$chan_entry.key();
+			let (_, err) = convert_chan_phase_err!($self, ChannelError::Close(
+				"Cannot update monitor for unfunded channels as they don't have monitors yet".into()),
+				$chan_entry.get_mut(), &channel_id);
+			$chan_entry.remove();
+			Err(err)
```
I think it leads to an early return and messes things up here.

dunxen commented:
> Whenever possible, if not using fixups, please keep rebases as their own push when also addressing feedback so that it's easier to tell what changed.

Sorry about that.
